### PR TITLE
Feat: First Pass at Pydantic

### PR DESCRIPTION
**Pull request by hatchet-temporary**
First pass at implementing Pydantic validation for workflow inputs and step outputs. For the API, I was thinking it'd make sense to let the user specify an `input_validator` at the workflow level and have the inputs be parsed with that, but I was having trouble figuring out a way to allow users to provide that and then using the same type as the return type of `workflow_input()`, so instead opted for allowing for an optional `validator` to `workflow_input` and letting the user `cast` (see the example). 

And then for step outputs, I was thinking it'd be nice to magically have type hints work (similar to FastAPI) for the individual steps - the negative here is there's no type support so you need to `cast` the `step_output` to the type you want in this case too. Not ideal, but maybe an 80% solution.

The example should be a good explanation of how the API should work!

### CODE REVIEW

**Review by abelanger5 — COMMENTED** [review]
> I was having trouble figuring out a way to allow users to provide that and then using the same type as the return type of workflow_input(), so instead opted for allowing for an optional validator to workflow_input

I'm not sure how to get type hinting to work with this, but I think this would be possible by storing a constructor for the `Context` along with the action (kind of like the action registry -- a mapping of actions to constructors). Or perhaps not a constructor, but something that transforms `Context -> Context`. So if an input validator is set on the workflow, we can call this function when initializing the step.

**Comment by hatchet-temporary** [comment]
> > I was having trouble figuring out a way to allow users to provide that and then using the same type as the return type of workflow_input(), so instead opted for allowing for an optional validator to workflow_input
> 
> I'm not sure how to get type hinting to work with this, but I think this would be possible by storing a constructor for the `Context` along with the action (kind of like the action registry -- a mapping of actions to constructors). Or perhaps not a constructor, but something that transforms `Context -> Context`. So if an input validator is set on the workflow, we can call this function when initializing the step.

Interesting idea! I'll try this out

File: hatchet_sdk/worker/action_listener_process.py [hunk 1]
```diff
@@ -87,15 +87,13 @@ async def start(self, retry_attempt=0):
...
-            self.listener: ActionListener = (
-                await self.dispatcher_client.get_action_listener(
-                    GetActionListenerRequest(
-                        worker_name=self.name,
-                        services=["default"],
-                        actions=self.actions,
-                        max_runs=self.max_runs,
-                        _labels=self.labels,
-                    )
+            self.listener = await self.dispatcher_client.get_action_listener(
```
Comment (hatchet-temporary) [inline hatchet_sdk/worker/action_listener_process.py:90]: just removed the type hint here - not needed anymore

File: examples/pydantic/worker.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+class ChildInput(BaseModel):
+    a: int
+    b: int
+
+
+class StepResponse(BaseModel):
+    status: str
+
+
+@hatchet.workflow()
```
Comment (abelanger5) [inline examples/pydantic/worker.py:49]: Should this workflow have an `input_validator` set?

**Review by abelanger5 — APPROVED** [review]

File: examples/pydantic/worker.py (hunk 2 above)
Comment (hatchet-temporary) [reply examples/pydantic/worker.py:49]: nice catch, thanks!

**Pull request merged**